### PR TITLE
Implement type family Elem

### LINKED ISSUE #170: Implement `Elem` type family
I found myself implementing it almost every time when using custom type errors. Probably a good idea to add it to the `Extra.Type` module :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by kutyel**
Resolves #170 

### HLint

- [ ] I've changed the exposed interface (add new reexports, remove reexports, rename reexported things, etc.).
  - [ ] I've updated [`hlint.dhall`](https://github.com/kowainik/relude/blob/master/hlint/hlint.dhall) accordingly to my changes (add new rules for the new imports, remove old ones, when they are outdated, etc.).
  - [ ] I've generated the new `.hlint.yaml` file (see [this instructions](https://github.com/kowainik/relude#generating-hlintyaml)).

### General

- [x] I've updated the [CHANGELOG](https://github.com/kowainik/relude/blob/master/CHANGELOG.md) with the short description of my latest changes.
- [x] All new and existing tests pass.
- [x] I keep the code style used in the files I've changed (see [style-guide](https://github.com/kowainik/org/blob/master/style-guide.md#haskell-style-guide) for more details).
- [x] I've used the [`stylish-haskell` file](https://github.com/kowainik/relude/blob/master/.stylish-haskell.yaml).
- [x] My change requires the documentation updates.
  - [x] I've updated the documentation accordingly.
- [ ] I've added the `[ci skip]` text to the docs-only related commit's name.

### CODE REVIEW

File: src/Relude/Extra/Type.hs [hunk 1]
```diff
@@ -114,3 +117,17 @@ Maybe (Snd (Int, Text)) :: *
 type family Snd (t :: k) :: k' where
     Snd '(_, y) = y
     Snd  (_, y) = y
+
+
+{- | Check that a type is an element of a list:
```
Comment (chshersh) [inline src/Relude/Extra/Type.hs]: Let's also add one more example with an empty list to show how it works on empty lists

File: src/Relude/Extra/Type.hs [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
-{-# LANGUAGE PolyKinds           #-}
-{-# LANGUAGE TypeApplications    #-}
-{-# LANGUAGE TypeFamilies        #-}
-{-# LANGUAGE TypeOperators       #-}
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE CPP                  #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE ExplicitNamespaces   #-}
+{-# LANGUAGE PolyKinds            #-}
+{-# LANGUAGE TypeApplications     #-}
```
Comment (vrom911) [inline src/Relude/Extra/Type.hs]: This is actually is already in the list of default extensions in the `.cabal` file, but I see that it was here before the PR...

File: src/Relude/Extra/Type.hs [hunk 3]
```diff
@@ -114,3 +117,17 @@ Maybe (Snd (Int, Text)) :: *
 type family Snd (t :: k) :: k' where
     Snd '(_, y) = y
     Snd  (_, y) = y
+
+
+{- | Check that a type is an element of a list:
+>>> :kind! (Elem Bool '[Int, Bool])
```
Comment (chshersh) [inline src/Relude/Extra/Type.hs]: Parenthesis here are redundant
```suggestion
>>> :kind! Elem Bool '[Int, Bool]
```

File: src/Relude/Extra/Type.hs [hunk 4]
```diff
@@ -114,3 +117,17 @@ Maybe (Snd (Int, Text)) :: *
...
+(Elem Bool '[Int, Bool]) :: Bool
+= 'True
+>>> :kind! (Elem String '[Int, Bool])
```
Comment (chshersh) [inline src/Relude/Extra/Type.hs]: And here :slightly_smiling_face: 
```suggestion
>>> :kind! Elem String '[Int, Bool]
```

File: src/Relude/Extra/Type.hs [hunk 5]
```diff
@@ -114,3 +117,17 @@ Maybe (Snd (Int, Text)) :: *
...
+(Elem String '[Int, Bool]) :: Bool
+= 'False
+-}
+type family Elem (e :: t) (es :: [t]) :: Bool where
+    Elem _ '[]       = 'False
+    Elem x (x ': xs) = 'True
+    Elem x (y ': xs) = Elem x xs
```
Comment (chshersh) [inline src/Relude/Extra/Type.hs]: I wonder, will it work if you write like this? :thinking: 
```suggestion
    Elem x (_ ': xs) = Elem x xs
```

Comment (kutyel) [reply src/Relude/Extra/Type.hs]: Yes, it works! 🎉

File: src/Relude/Extra/Type.hs (hunk 2 above)
Comment (kutyel) [reply src/Relude/Extra/Type.hs]: Yes, apparently CI needs it because removing it made the build to fail 😓

**Comment by chshersh** [comment]
Nice work!

**Pull request merged**